### PR TITLE
Add window centering.

### PR DESCRIPTION
**Pull request by mralve**
Added window centering on windows and linux (x11) at least.

Added convenient functions to be able to center the window quickly, also made it default on windows, but this is togglable.

- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ - ] Created an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/tomaka/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by mralve** [comment]
Is currently testing/working on Linux

**Comment by vberger** [comment]
It looks like your PR changes a lot of unrelated code, I suspect you've run rustfmt on the files?

Could you avoid doing so? It makes the pull request pretty difficult to review.

**Comment by mralve** [comment]
@vberger yeah exactly sorry, it's my vscode having auto fmt on, I will see if I can correct it in a new commit.

**Comment by mralve** [comment]
@vberger Now this should be ready for review, although I don't possess the tools for mac so it has to be done by someone else. ^_^

File: examples/window.rs [hunk 1]
```diff
@@ -8,6 +8,8 @@ fn main() {
         .build(&events_loop)
         .unwrap();
 
+    _window.center();
```
Comment (ZeGentzy) [inline examples/window.rs]: Nit: remove the underscore.

File: examples/window_icon.rs [hunk 2]
```diff
@@ -27,8 +27,11 @@ fn main() {
         // At present, this only does anything on Windows and X11, so if you want to save load
         // time, you can put icon loading behind a function that returns `None` on other platforms.
         .with_window_icon(Some(icon))
+        .with_centering(true)
         .build(&events_loop)
         .unwrap();
+    
+    //window.center();
```
Comment (ZeGentzy) [inline examples/window_icon.rs]: Nit: either uncomment, or remove.

File: src/lib.rs [hunk 3]
```diff
@@ -510,6 +510,16 @@ pub struct WindowAttributes {
     /// The default is `false`.
     pub always_on_top: bool,
 
+    /// Whether the window will be centered on screen.
+    ///
+    /// The default is true.
```
Comment (ZeGentzy) [inline src/lib.rs]: true -> \`true\`

File: src/lib.rs [hunk 4]
```diff
@@ -510,6 +510,16 @@ pub struct WindowAttributes {
...
+    pub center_window: bool,
+
+    /// Decides what monitor the window will start on. Users primary monitor is -1.
```
Comment (ZeGentzy) [inline src/lib.rs]: what -> which
Users -> The user's
-1 -> \`-1`

File: src/lib.rs [hunk 5]
```diff
@@ -510,6 +510,16 @@ pub struct WindowAttributes {
...
+    ///
+    /// The default is -1,
```
Comment (ZeGentzy) [inline src/lib.rs]: -1 -> \`-1`

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,6 @@
 # Unreleased
-
+- Added in WindowAttributes `center_window: bool` `start_monitor: i16`, Window centering is by default enabled on windows.
```
Comment (ZeGentzy) [inline CHANGELOG.md]: "Added to \`WindowAttributes\` the attributes  \`center_window: bool\` and \`start_monitor: i16`. Window centering is by default enabled on windows."

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,6 @@
...
+- Added convenience functions to manage window centering, `center()` ,`set_center_monitor()`, `set_centered()` see more in window struct.
```
Comment (ZeGentzy) [inline CHANGELOG.md]: Change to this instead:
"Added the following convenience functions for managing window centering: `center()` ,`set_center_monitor()`, and `set_centered()`. Refer to the documentation for how to use them."

File: src/platform/windows/window.rs [hunk 8]
```diff
@@ -715,15 +715,43 @@ unsafe fn init(
...
 
     if let Some(dimensions) = attributes.dimensions {
         win.set_inner_size(dimensions);
     }
 
+    if attributes.center_window && !&attributes.fullscreen.is_some(){
+        let window_size = win.get_outer_size();
+
+        let monitor;
+        if attributes.start_monitor == -1 {
```
Comment (ZeGentzy) [inline src/platform/windows/window.rs]: Why is `start_monitor` an `i16`? This should be an `Option<MonitorId>` instead.

File: src/window.rs [hunk 9]
```diff
@@ -380,6 +395,39 @@ impl Window {
...
     }
+        /// Centers the window on the current monitor the window lives in.
+    pub fn center(&self) {
+        self.set_centered(self.get_current_monitor())
+    }
+
+    /// Centers the window on the specified target monitor, falls back to primary monitor if out range.
+    pub fn set_center_monitor(&self, monitor_target: usize) {
+        if monitor_target < self.get_available_monitors().count() {
+            self.set_centered(self.get_available_monitors().nth(monitor_target).unwrap())
```
Comment (ZeGentzy) [inline src/window.rs]: This should be accepting a `Option<MonitorId>`, not usize. 
`self.set_centered(monitor_target.unwrap_or_else(|_| self.get_primary_monitor()))`

File: src/window.rs [hunk 10]
```diff
@@ -69,6 +69,21 @@ impl WindowBuilder {
         self
     }
 
+        /// Sets so the window will start centered on screen.
```
Comment (ZeGentzy) [inline src/window.rs]: Why the weird formatting?
Also, "Sets so" -> "Sets whether or not"

File: src/window.rs [hunk 11]
```diff
@@ -69,6 +69,21 @@ impl WindowBuilder {
...
+    #[inline]
+    pub fn with_centering(mut self, centered: bool) -> WindowBuilder {
+        self.window.center_window = centered;
+        self
+    }
+
+    /// Sets what monitor the window will start in.
```
Comment (ZeGentzy) [inline src/window.rs]: "what" -> "which"

File: src/window.rs [hunk 12]
```diff
@@ -69,6 +69,21 @@ impl WindowBuilder {
...
+    /// where -1 is the users primary monitor.
```
Comment (ZeGentzy) [inline src/window.rs]: -1 -> \`-1`

File: src/window.rs (hunk 11 above)
Comment (ZeGentzy) [inline src/window.rs]: Turn the period at the end to a comma.

File: src/window.rs (hunk 12 above)
Comment (ZeGentzy) [inline src/window.rs]: users -> user's

File: src/platform/linux/x11/window.rs [hunk 13]
```diff
@@ -380,6 +380,35 @@ impl UnownedWindow {
                 window.set_always_on_top_inner(window_attrs.always_on_top).queue();
             }
 
+            if window_attrs.center_window && !window_attrs.fullscreen.is_some() 
+            {
+                let window_size = window.get_outer_size();
+
+                let monitor;
+                if window_attrs.start_monitor == -1 {
```
Comment (ZeGentzy) [inline src/platform/linux/x11/window.rs]: As mentioned, this code would look better if you just stored a `MonitorId` here.

File: src/window.rs (hunk 9 above)
Comment (mralve) [reply src/window.rs]: well, the goal with set_center_monitor() is to only supply a simple monitor index, and if that is out of range then it falls back to the primary monitor, set_center_monitor is better renamed to set_target_monitor() and monitor_target > monitor_index to reflect that idea better. otherwise, if you want to use a MonitorID then simply use set_centered() instead.

File: src/platform/windows/window.rs (hunk 8 above)
Comment (mralve) [reply src/platform/windows/window.rs]: This is done for simplicity. only supplying a monitors index in display config is better IMO, start_monitor is renamed to monitor_index to better reflect that.

Comment (vberger) [reply src/platform/windows/window.rs]: We don't really have any guarantee that the order will be the same across starts of the application, so configuring apps by monitor index cannot work reliably. Integrating such an API in winit would be a false promise that this would work.

File: FEATURES.md [hunk 14]
```diff
@@ -88,7 +88,7 @@ If your PR makes notable changes to Winit's features, please update this section
 - **Popup / modal windows**: Windows can be created relative to the client area of other windows, and parent
   windows can be disabled in favor of popup windows. This feature also guarantees that popup windows
   get drawn above their owner.
-
+- **Window centering**: Sets the window position to be in the center of the given monitor.
```
Comment (vberger) [inline FEATURES.md]: If you add an entry here, you need to also add it to the compatibility matrix below.

File: src/window.rs [hunk 15]
```diff
@@ -380,6 +395,40 @@ impl Window {
     pub fn set_decorations(&self, decorations: bool) {
         self.window.set_decorations(decorations)
     }
+    
+    /// Centers the window on the current monitor the window lives in.
```
Comment (vberger) [inline src/window.rs]: Please specify the platform-dependent behavior of this function. For example it cannot work at all on Wayland, as the app cannot control its placement on the monitor.

File: src/window.rs [hunk 16]
```diff
@@ -380,6 +395,40 @@ impl Window {
...
+    pub fn set_target_monitor(&self, monitor_index: usize) {
+        if monitor_index < self.get_available_monitors().count() {
+            self.set_centered(self.get_available_monitors().nth(monitor_index).unwrap())
+        } else {
+            self.set_centered(self.get_primary_monitor());
+        }
+    }
+
+    /// Sets the window position to be in the center of the given monitor.
+    /// Will do nothing if the window is in fullscreen.
```
Comment (vberger) [inline src/window.rs]: Same here, this method behavior is platform dependent.

File: src/window.rs [hunk 17]
```diff
@@ -380,6 +395,40 @@ impl Window {
...
+        let monitor_size = monitor.get_dimensions();
+        let monitor_position = monitor.get_position();
+
+        let mut monitor_window_position: LogicalPosition = (0.0, 0.0).into();
+        monitor_window_position.x =
+            monitor_position.x + (monitor_size.width * 0.5) - (&window_size.width * 0.5);
+        monitor_window_position.y =
+            monitor_position.y + (monitor_size.height * 0.5) - (&window_size.height * 0.5);
+
+        &self.set_position(monitor_window_position);
```
Comment (vberger) [inline src/window.rs]:
```suggestion
        self.set_position(monitor_window_position);
```

File: src/platform/windows/window.rs (hunk 8 above)
Comment (mralve) [reply src/platform/windows/window.rs]: Ahh, I see, well then it's better to drop the set_center_monitor() function and use the Option<MonitorId> directly in the display config?

File: src/window.rs (hunk 9 above)
Comment (mralve) [reply src/window.rs]: dropping set_center_monitor() as set_centered() already do this.

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,5 +1,6 @@
 # Unreleased
-
+- Added to `WindowAttributes` the attributes `center_window: bool` and `start_monitor: i16`. Window centering is by default enabled.
+- Added the following convenience functions for managing window centering: center(), and set_centered(). Refer to the documentation for how to use them.
```
Comment (ZeGentzy) [inline CHANGELOG.md]: Oh, I forgot. We should surround the two functions with backticks (`)

File: src/platform/linux/x11/window.rs [hunk 19]
```diff
@@ -380,6 +380,34 @@ impl UnownedWindow {
...
             }
 
+            if window_attrs.center_window && !window_attrs.fullscreen.is_some() 
+            {
+                let window_size = window.get_outer_size();
+
+                let monitor_size;
+                let monitor_position;
+
+                if window_attrs.start_monitor.is_none() {
```
Comment (ZeGentzy) [inline src/platform/linux/x11/window.rs]: Haven't tested if this compiles, but how about something like this?
```
let monitor = match window_attrs.start_monitor {
    None => window.get_primary_monitor(),
    Some(m) => m.inner,
};

monitor_size = monitor.get_dimensions();
monitor_position = monitor.get_position();
```

File: src/platform/linux/x11/window.rs [hunk 20]
```diff
@@ -380,6 +380,34 @@ impl UnownedWindow {
...
+                    monitor_size = monitor.get_dimensions();
+                    monitor_position = monitor.get_position();
+                }
+
+
+                let mut monitor_window_position: LogicalPosition = (0.0, 0.0).into();
+                monitor_window_position.x =
+                    monitor_position.x + (monitor_size.width * 0.5) - (&window_size.unwrap().width * 0.5);
+                monitor_window_position.y =
+                    monitor_position.y + (monitor_size.height * 0.5) - (&window_size.unwrap().height * 0.5);
```
Comment (ZeGentzy) [inline src/platform/linux/x11/window.rs]: What if `monitor_position.{x,y}` != 0? Won't this not centre the window?

File: src/platform/windows/window.rs [hunk 21]
```diff
@@ -715,15 +715,39 @@ unsafe fn init(
...
     }
 
     if let Some(dimensions) = attributes.dimensions {
         win.set_inner_size(dimensions);
     }
 
+    if attributes.center_window && !&attributes.fullscreen.is_some(){
+        let window_size = win.get_outer_size();
+
+        let monitor;
```
Comment (ZeGentzy) [inline src/platform/windows/window.rs]: Haven't tested if this compiles, but how about something like this?
```
let monitor = match window_attrs.start_monitor {
    None => window.get_primary_monitor(),
    Some(m) => m.inner,
};
```

File: src/platform/windows/window.rs [hunk 22]
```diff
@@ -715,15 +715,39 @@ unsafe fn init(
...
+        } else {
+            monitor = attributes.start_monitor.unwrap().inner;
+        }
+
+        let monitor_size = monitor.get_dimensions();
+        let monitor_position = monitor.get_position();
+
+
+        let mut monitor_window_position: LogicalPosition = (0.0, 0.0).into();
+        monitor_window_position.x =
```
Comment (ZeGentzy) [inline src/platform/windows/window.rs]: What if `monitor_position.{x,y}` != 0? Won't this not centre the window?

File: src/window.rs [hunk 23]
```diff
@@ -380,6 +395,45 @@ impl Window {
     pub fn set_decorations(&self, decorations: bool) {
         self.window.set_decorations(decorations)
     }
+    
+    /// Centers the window on the current monitor the window lives in.
+    /// 
+    /// ## Platform-specific
+    /// 
+    /// This only works on Windows and X11.
+    /// it's not tested on macOS.
```
Comment (ZeGentzy) [inline src/window.rs]: You mean "It's not implemented on MacOS"? I don't see a MacOS impl anywhere.

At minimum, "it's" -> "It's"

File: src/window.rs [hunk 24]
```diff
@@ -380,6 +395,45 @@ impl Window {
...
+        self.set_centered(self.get_current_monitor())
+    }
+
+    /// Sets the window position to be in the center of the given monitor.
+    /// Will do nothing if the window is in fullscreen.
+    /// 
+    /// ## Platform-specific
+    /// 
+    /// This only works on Windows and X11.
+    /// it's not tested on macOS.
```
Comment (ZeGentzy) [inline src/window.rs]: Same as above.

File: src/window.rs [hunk 25]
```diff
@@ -380,6 +395,45 @@ impl Window {
...
+        };
+
+        let monitor_size = monitor.get_dimensions();
+        let monitor_position = monitor.get_position();
+
+        let mut monitor_window_position: LogicalPosition = (0.0, 0.0).into();
+        monitor_window_position.x =
+            monitor_position.x + (monitor_size.width * 0.5) - (&window_size.width * 0.5);
+        monitor_window_position.y =
+            monitor_position.y + (monitor_size.height * 0.5) - (&window_size.height * 0.5);
```
Comment (ZeGentzy) [inline src/window.rs]: Same as above.

File: src/window.rs (hunk 23 above)
Comment (ZeGentzy) [reply src/window.rs]: Nvm, I guess the default impls bellow act as MacOS's.

Anyways, "it's" -> "It's", "macOS" -> "MacOS"

File: src/platform/linux/x11/window.rs (hunk 20 above)
Comment (mralve) [reply src/platform/linux/x11/window.rs]: I don't see any problem here, it centers on both of my monitors. have tried entering custom values to test, but that works as expected.

File: src/platform/windows/window.rs (hunk 21 above)
Comment (mralve) [reply src/platform/windows/window.rs]: ahh right, matching is so nice, I'm just too used to write ifs, coming from c#

File: src/platform/linux/x11/window.rs (hunk 20 above)
Comment (ZeGentzy) [reply src/platform/linux/x11/window.rs]: Ah, I thought `monitor_position` was the window's position on the monitor. Looking at the code again, I realize I was mistaken.

File: src/lib.rs [hunk 26]
```diff
@@ -531,6 +541,8 @@ impl Default for WindowAttributes {
             title: "winit window".to_owned(),
             maximized: false,
             fullscreen: None,
+            center_window: true,
```
Comment (aloucks) [inline src/lib.rs]: I don't think it's right to set `center_window: true` as the default behavior. Let the platform decide where to put it. This is better off as opt-in behavior.

File: src/lib.rs [hunk 27]
```diff
@@ -510,6 +510,16 @@ pub struct WindowAttributes {
...
 
+    /// Whether the window will be centered on screen.
+    ///
+    /// The default is `true`.
+    pub center_window: bool,
+
+    /// Decides which monitor the window will start on. The user's primary monitor is `None`.
+    ///
+    /// The default is `None`,
+    pub start_monitor: Option<MonitorId>,
```
Comment (aloucks) [inline src/lib.rs]: I don't think it's unreasonable to require the user to select the monitor first (e.g. `EventsLoop::get_primary_monitor()`).

Why not remove `start_monitor` and only have:

```rust
center_window: Option<MonitorId>
```

File: src/window.rs [hunk 28]
```diff
@@ -380,6 +395,45 @@ impl Window {
...
+    
+    /// Centers the window on the current monitor the window lives in.
+    /// 
+    /// ## Platform-specific
+    /// 
+    /// This only works on Windows and X11.
+    /// It's not tested on macOS.
+    /// 
+    /// This has no effect on Android, Wayland or iOS.
+    pub fn center(&self) {
```
Comment (aloucks) [inline src/window.rs]: Why is this needed in addition to `set_centered`? The naming of state changing functions is already inconsistent (see also #728). I think having multiple ways to do the same thing is probably not a great idea, especially when getting the default monitor is already really easy.

File: src/window.rs [hunk 29]
```diff
@@ -380,6 +395,45 @@ impl Window {
     pub fn set_decorations(&self, decorations: bool) {
         self.window.set_decorations(decorations)
     }
+    
+    /// Centers the window on the current monitor the window lives in.
+    /// 
+    /// ## Platform-specific
+    /// 
+    /// This only works on Windows and X11.
+    /// It's not tested on macOS.
```
Comment (Osspial) [inline src/window.rs]: It's misleading to say that this isn't tested on macOS - it just isn't implemented.

File: src/window.rs [hunk 30]
```diff
@@ -380,6 +395,45 @@ impl Window {
...
+        self.set_centered(self.get_current_monitor())
+    }
+
+    /// Sets the window position to be in the center of the given monitor.
+    /// Will do nothing if the window is in fullscreen.
+    /// 
+    /// ## Platform-specific
+    /// 
+    /// This only works on Windows and X11.
+    /// It's not tested on macOS.
```
Comment (Osspial) [inline src/window.rs]: See previous comment on macOS.

File: FEATURES.md [hunk 31]
```diff
@@ -154,6 +154,7 @@ Legend:
 |Window transparency              |✔️     |✔️     |✔️         |✔️             |**N/A**|**N/A**|**N/A**   |
 |Window maximization              |✔️     |✔️     |✔️         |✔️             |**N/A**|**N/A**|**N/A**   |
 |Window maximization toggle       |✔️     |✔️     |✔️         |✔️             |**N/A**|**N/A**|**N/A**   |
+|Window centering                 |✔️     |❌     |✔️         |❌             |**N/A**|**N/A**|**N/A**   |
```
Comment (Osspial) [inline FEATURES.md]: This should *probably* be **N/A** on Wayland, since it doesn't seem like the sort of thing Wayland would support. cc @vberger

File: CHANGELOG.md [hunk 32]
```diff
@@ -1,5 +1,6 @@
 # Unreleased
-
+- Added to `WindowAttributes` the attributes `center_window: bool` and `start_monitor: i16`. Window centering is by default enabled.
```
Comment (Osspial) [inline CHANGELOG.md]: `start_monitor` is `Option<MonitorId>`, not `i16`. Window centering should default to being disabled.

**Comment by Osspial** [comment]
On a higher level, I'd like to ask: why is this PR necessary? The [`set_centered` method's implementation](https://github.com/rust-windowing/winit/blob/e1edc1d6b7e42b8d7632b18dfd87f5c75b831e63/src/window.rs#L420-L436) shows that it's entirely possible to implement this without access to any of Winit's internals, and bringing this into our codebase increases the codebase's complexity. Admittedly, It's not a substantial increase in this case, but it sets an uncomfortable prescient for pulling in code that's *convenient* but that can be implemented entirely outside of Winit.

**Comment by mralve** [comment]
@Osspial Whilst you bring up great points I would understand if you reject this pr, But I would argue that for a "window creation and management library." as stated in the readme, a library that provides basic functionality for window creation and management, being able to control the window starting monitor, and if it's going to be centered on the screen out of the box is seen as rather basic IMO, while yes being convenient i would see this as clearing up unnecessary redundancy in the view from multiple projects needing this trivial functionality.

**Comment by Osspial** [comment]
So, it seems like I was lacking a bit of nuance with that initial response! To clarify myself: I'm entirely okay with adding a field in `WindowAttributes` to set the window's default monitor. I just don't like bringing `set_centered` into Winit's codebase, since it can be entirely implemented with publicly-visible, non-platform-specific API calls.

Also, given the above I'd rather refer to this feature as *default monitor* rather than *window centering*, since that exposes looser API guarantees and opens the door for the API to be implemented on platforms that have multiple displays but not necessarily a desktop surface.

**Comment by mralve** [comment]
Great having a default monitor property, would kinda require window centering by default unless there is better a way to give the os responsibility to place the window on the right screen and by its deafult window position?

File: src/platform/windows/window.rs [hunk 33]
```diff
@@ -715,15 +715,37 @@ unsafe fn init(
...
+    if let Some(_) = &attributes.fullscreen {
+        win.set_fullscreen(attributes.fullscreen.clone());
         force_window_active(win.window.0);
     }
 
     if let Some(dimensions) = attributes.dimensions {
         win.set_inner_size(dimensions);
     }
 
+    if attributes.center_window.is_some() && !&attributes.fullscreen.is_some(){
```
Comment (ZeGentzy) [inline src/platform/windows/window.rs]: We can replace all this code with something along the lines of this, right (pseudo code):
```
if attributes.fullscreen.is_none() {
    if let Some(mon) = attributes.center_window {
        win.set_centered(mon);
    }
}
```

File: src/platform/linux/x11/window.rs [hunk 34]
```diff
@@ -380,6 +380,27 @@ impl UnownedWindow {
                 window.set_always_on_top_inner(window_attrs.always_on_top).queue();
             }
 
+            if window_attrs.center_window.is_some() && !window_attrs.fullscreen.is_some()
```
Comment (ZeGentzy) [inline src/platform/linux/x11/window.rs]: https://github.com/rust-windowing/winit/pull/857/files#r286254032

**Comment by ZeGentzy** [comment]
I'll chip in: maybe instead of `set_centered` we could rename it to `set_monitor`, and instead of `center_window`, `default_monitor`, as suggested. 

The new function will by default try to place the window in the center. If that's not possible, just anywhere on the monitor will do. Not sure if the latter is possible on Wayland, but if it's not, I guess it can be a no-op.

**Comment by ZeGentzy** [comment]
Closing for inactivity.
